src/layouts.jl

```diff
@@ -0,0 +1,139 @@
+
+"""
+    StrideLayout(A)
+
+Produces an array whose elements correspond to the linear buffer position of `A`'s elements.
+"""
+struct StrideLayout{N,R,O1,S,A<:Tuple{Vararg{Any,N}}} <: AbstractArray2{Int,N}
+    rank::R
+    offset1::O1
+    strides::S
+    axes::A
+end
+
+offset1(x::StrideLayout) = getfield(x, :offset1)
+offsets(x::StrideLayout) = map(static_first, axes(x))
+axes(x::StrideLayout) = getfield(x, :axes)
+@inline function axes(x::StrideLayout{N}, i::Int) where {N}
+    if i > N
+        return static(1):1
+    else
+        return getfield(getfield(x, :axes), i)
+    end
+end
+@inline function axes(x::StrideLayout{N}, ::StaticInt{i}) where {N,i}
+    if i > N
+        return static(1):static(1)
+    else
+        return getfield(getfield(x, :axes), i)
+    end
+end
+strides(x::StrideLayout) = getfield(x, :strides)
+stride_rank(x::StrideLayout) = getfield(x, :rank)
+
+@inline function StrideLayout(x::DenseArray)
+    a = axes(x)
+    return StrideLayout(
+        stride_rank(x),
+        offset1(x),
+        size_to_strides(map(static_length, a), static(1)),
+        a
+    )
+end
+
+# TODO optimize this
+@inline function StrideLayout(x)
+    return StrideLayout(
+        stride_rank(x),
+        offset1(x),
+        strides(x),
+        axes(x)
+    )
+end
+
+##############
+### layout ###
+##############
+layout(x, i) = layout(x)
+layout(x, i::AbstractVector{<:Integer}) = _maybe_linear_layout(IndexStyle(x), x)
+layout(x, i::Integer) = _maybe_linear_layout(IndexStyle(x), x)
+layout(x, i::AbstractCartesianIndex{1}) = _maybe_linear_layout(IndexStyle(x), x)
+function layout(x, i::AbstractVector{AbstractCartesianIndex{1}})
+    return _maybe_linear_layout(IndexStyle(x), x)
+end
+_maybe_linear_layout(::IndexLinear, x) = _as_index(eachindex(x))
+_maybe_linear_layout(::IndexStyle, x) = layout(x)
+layout(x::StrideLayout) = x
+layout(x::LinearIndices) = x
+layout(x::CartesianIndices) = x
+function layout(x)
+    if defines_strides(x)
+        return StrideLayout(x)
+    else
+        return _layout_indices(IndexStyle(x), axes(x))
+    end
+end
+function layout(x::Transpose)
+    if defines_strides(x)
+        return StrideLayout(x)
+    else
+        return Transpose(layout(parent(x)))
+    end
+end
+function layout(x::Adjoint{T}) where {T<:Number}
+    if defines_strides(x)
+        return StrideLayout(x)
+    else
+        return Transpose(layout(parent(x)))
+    end
+end
+function layout(x::PermutedDimsArray{T,N,perm,iperm}) where {T,N,perm,iperm}
+    if defines_strides(x)
+        return StrideLayout(x)
+    else
+        p = layout(parent(x))
+        return PermutedDimsArray{eltype(p),ndims(p),perm, iperm,typeof(p)}(p)
+    end
+end
+function layout(x::SubArray)
+    if defines_strides(x)
+        return StrideLayout(x)
+    else
+        return @inbounds(view(layout(parent(x)), x.indices...))
+    end
+end
+_layout_indices(::IndexStyle, axs) = CartesianIndices(axs)
+_layout_indices(::IndexLinear, axs) = LinearIndices(axs)
```

Comment:
Whether to use linear or Cartesian indexing is of course context dependent (in the rewrite, I'll add support for switching between representations).
E.g., if you're summing a `Matrix{Float64}` or a `Adjoint{Float64,Matrix{Float64}}`, you'd want to use linear indexing in both cases (for the `Adjoint`, the linear indices would be traversing it in row major order). Same if you're calculating the sum of 2 instances of `Matrix{Float64}` or 2 instances of `Adjoint{Float64,Matrix{Float64}}`. However, if you sum add a `Matrix{Float64}` to a `Adjoint{Float64,Matrix{Float64}}`, you suddenly need cartesian indexing.
`eachindex` gets some of this right, but it'll always return `IndexCartesian()` for the `Adjoint`, meaning performance for those sums/additions is worse than necessary.

Is there a "combine layouts"?

I'll get LoopVectorization to be able to transform between these correctly in the rewrite.
One of my goals there is for something like this to work:
```julia
@avx begin
  C .= 0
  for n in indices((C,B),2), m in indices((C,A),1), for k in indices((A,B),(2,1))
    C[m,n] += A[m,k] * B[k,n]
  end
end
```
and to have LoopVectorization be capable of converting the linear indexing of `C .= 0` into cartesian indexing, and then fuse that broadcast with the following loop.

Reply:
What you're describing is basically what I was getting at with the reference to MLIR. The final "layout" after code gen would be the result of scoping (for summation this would be unordered access to all elements once) which informs how to combine each layer of the arrays.